### PR TITLE
Validate VK_NV_cooperative_matrix

### PR DESCRIPTION
**Pull request by jeffbolznv**

### CODE REVIEW

**Comment by jeffbolznv** [comment]
Can somebody review this? @mikes-lunarg @jzulauf-lunarg ?

File: layers/core_validation_error_enums.h [hunk 1]
```diff
@@ -181,6 +181,9 @@ static const char DECORATE_UNUSED *kVUID_Core_Shader_PushConstantNotAccessibleFr
 static const char DECORATE_UNUSED *kVUID_Core_Shader_PushConstantOutOfRange = "UNASSIGNED-CoreValidation-Shader-PushConstantOutOfRange";
 static const char DECORATE_UNUSED *kVUID_Core_Shader_MissingPointSizeBuiltIn = "UNASSIGNED-CoreValidation-Shader-PointSizeMissing";
 static const char DECORATE_UNUSED *kVUID_Core_Shader_PointSizeBuiltInOverSpecified = "UNASSIGNED-CoreValidation-Shader-PointSizeOverSpecified";
+static const char DECORATE_UNUSED *kVUID_Core_Shader_CooperativeMatrixSupportedStages = "UNASSIGNED-CoreValidation-Shader-CooperativeMatrixSupportedStages";
```
Comment (jzulauf-lunarg) [inline layers/core_validation_error_enums.h]: This seems like missing VU's for your extension.  There should be an issue or an MR with the working group to add them, no?

File: layers/shader_validation.cpp [hunk 2]
```diff
@@ -1827,6 +1828,284 @@ static bool ValidateShaderStageInputOutputLimits(layer_data *dev_data, shader_mo
...
+                    return VK_COMPONENT_TYPE_MAX_ENUM_NV;
+            }
+        default:
+            return VK_COMPONENT_TYPE_MAX_ENUM_NV;
+    }
+}
+
+// Validate SPV_NV_cooperative_matrix behavior that can't be statically validated
+// in SPIRV-Tools (e.g. due to specialization constant usage).
+static bool ValidateCooperativeMatrix(layer_data *dev_data, shader_module const *src, VkPipelineShaderStageCreateInfo const *pStage,
```
Comment (jzulauf-lunarg) [inline layers/shader_validation.cpp]: I'm not seeing a Validation check for -- I've got CooperativeMatrix stuff in my shader, but it hasn't been properly enabled....

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -2562,6 +2567,25 @@ void PostCallRecordCreateDevice(VkPhysicalDevice gpu, const VkDeviceCreateInfo *
...
+    if (device_data->extensions.vk_nv_cooperative_matrix) {
+        // Get the needed cooperative_matrix properties
+        auto cooperative_matrix_props = lvl_init_struct<VkPhysicalDeviceCooperativeMatrixPropertiesNV>();
+        auto prop2 = lvl_init_struct<VkPhysicalDeviceProperties2KHR>(&cooperative_matrix_props);
+        instance_data->dispatch_table.GetPhysicalDeviceProperties2KHR(gpu, &prop2);
+        device_data->phys_dev_ext_props.cooperative_matrix_props = cooperative_matrix_props;
+
+        uint32_t numCooperativeMatrixProperties = 0;
+        instance_data->dispatch_table.GetPhysicalDeviceCooperativeMatrixPropertiesNV(gpu, &numCooperativeMatrixProperties, NULL);
+        device_data->cooperative_matrix_properties.resize(numCooperativeMatrixProperties);
```
[the remark this answers is not in the document]
Comment (jeffbolznv) [reply layers/core_validation.cpp]: Nice, done.

File: layers/core_validation.cpp [hunk 4]
```diff
@@ -2562,6 +2567,25 @@ void PostCallRecordCreateDevice(VkPhysicalDevice gpu, const VkDeviceCreateInfo *
...
+
+        for (uint32_t i = 0; i < numCooperativeMatrixProperties; ++i) {
+            device_data->cooperative_matrix_properties[i].sType = VK_STRUCTURE_TYPE_COOPERATIVE_MATRIX_PROPERTIES_NV;
+            device_data->cooperative_matrix_properties[i].pNext = NULL;
+        }
+
+        instance_data->dispatch_table.GetPhysicalDeviceCooperativeMatrixPropertiesNV(
+            gpu, &numCooperativeMatrixProperties, &device_data->cooperative_matrix_properties[0]);
```
[the remark this answers is not in the document]
Comment (jeffbolznv) [reply layers/core_validation.cpp]: done

File: layers/shader_validation.cpp (hunk 2 above)
Comment (jeffbolznv) [reply layers/shader_validation.cpp]: Added (and this also addresses the other question about querying the Features but not using them).

File: layers/core_validation_error_enums.h (hunk 1 above)
Comment (jeffbolznv) [reply layers/core_validation_error_enums.h]: These are rules in the spirvenv appendix, which doesn't get VUIDs assigned. I'll file a vulkan spec bug about this.

**Comment by jzulauf-lunarg** [comment]
@greg-lunarg  -- if you could focus on the shader_validation.cpp piece, if not, tag ChrisF?

**Comment by jeffbolznv** [comment]
Ping for an update. @greg-lunarg?

**Comment by jeffbolznv** [comment]
Rebased again. This has been sitting for over a month. @mikes-lunarg can it be merged?

**Comment by jeffbolznv** [comment]
Ping again for review/merge. @mikes-lunarg?

**Pull request merged**